From a18b271f486fc800ca8f1cf908025d359b3684cd Mon Sep 17 00:00:00 2001
From: Steffen Schulz <steffen.schulz@intel.com>
Date: Sat, 28 Sep 2019 02:09:38 +0200
Subject: [PATCH 03/11] kAFL v2 frontend requires us to start counting at 0

---
 pt.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pt.c b/pt.c
index b3088665..6e11c22e 100644
--- a/pt.c
+++ b/pt.c
@@ -78,7 +78,7 @@ void pt_setup_bitmap(void* ptr){
 void pt_reset_bitmap(void){
 	if(bitmap){
 		last_ip = 0ULL;
-		memset(bitmap, 0xff, kafl_bitmap_size);
+		memset(bitmap, 0x00, kafl_bitmap_size);
 	}
 }
 
-- 
2.26.2

